회원 삭제 soft delete로 변경

SUBIN606 opened the pull request:
회원 삭제 시 데이터를 실제로 삭제하지 않고, 삭제 상태를 변경하는 soft delete 방식으로 수정했습니다.

회원 탈퇴 후 바로 데이터를 삭제해 버리면,
- 배송 중인 고양이 장난감 상품의 출고 처리 문제
- 탈퇴 후 일정기간 동안 재가입 방지 불가

의 문제점을 발견했기 때문입니다.

`User` 엔티티 클래스에 `deleted` 필드를 추가해 삭제 상태를 나타내고, 삭제 상태로 변경되게 하는 `destroy` 메서드도 추가했습니다.
자세한 사용 방법은 테스트 코드를 참고해 주시기 바랍니다.

SUBIN606 committed:
UserTest 회원 삭제 테스트 추가

- 회원 삭제를 soft delete로 변경함에 따라 destroy 메서드 테스트를 추가합니다.

SUBIN606 committed:
회원 삭제 방식을 soft delete로 변경

- User 엔티티에 삭제 여부 상태를 나타내는 deleted 필드를 추가합니다.
- 회원 삭제 상태를 true로 변경하는 destroy 메서드를 추가합니다.

SUBIN606 committed:
회원 삭제 서비스 테스트 수정

- 회원 삭제 후 deleted 값 변경 여부를 검증합니다.

SUBIN606 committed:
회원 삭제 서비스, 회원 엔티티 변경

- UserRepository에 탈퇴하지 않은 회원만 조회하기 위해 findByIdAndDeletedIsFalse 메서드를 추가합니다.
- User 엔티티의 deleted 컬럼에 insertable 설정을 false로 변경해 default value로 저장되게 합니다.

SUBIN606 committed:
회원 삭제 요청 테스트 수정

SUBIN606 committed:
회원 리포지토리 메서드 명 findByEmail -> findByEmailAndDeletedIsFalse 변경

- 회원 로그인 시 사용하는 findByEmail 메서드를 탈퇴한 회원의 로그인을 막기 위해
findByEmailAndDeletedIsFalse로 변경합니다.

johngrib reviewed (commented):
PR 타이틀에서 `[week7]`은 없어도 괜찮을 것 같아요. 코드숨이 아니라 회사라고 생각하고 작업한다고 상상해보면 도움이 될 거에요.

![image](https://user-images.githubusercontent.com/1855714/167129629-eaa72420-2c74-477a-907d-b70b86855994.png)

그리고 "강의 내용 대로"라는 말도 없어도 괜찮을 것 같습니다.

![image](https://user-images.githubusercontent.com/1855714/167129795-05def0b6-2687-4e3c-8ff8-71c5d20db2f1.png)

PR 본문은 이렇게 작성해 보세요. `what` - `why` - `how` 를 떠올리며 작성하는 것입니다.

다음 예제는 제가 방금 작성한 것입니다. 읽고 다음 PR 본문을 어떻게 작성할지 궁리해 보세요.

```
회원을 삭제할 때 데이터를 실제로 삭제하지 않고, 삭제 상태만 변경하도록 수정했습니다.

회원이 탈퇴했을 때 회원 데이터를 곧바로 삭제하게 되면, 배송중인 고양이 장난감 상품의 출고 처리에 문제가 생기는 것을 발견했기 때문입니다.

이 작업을 위해 User 엔티티에 `deleted`라는 필드를 추가하고, `destroy` 메소드도 추가했습니다.

자세한 사용방법은 테스트 코드를 읽어보시면 됩니다.
```

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/users/UserTest.java:
```diff
@@ -127,5 +127,18 @@ void it_init_password_field() {
+    @Nested
+    class Describe_destroy {
+
+        @DisplayName("deleted 필드를 true로 변경한다.")
```
`deleted` 필드는 private 이므로 이 테스트의 주요 인물로 등장하지 않아도 괜찮습니다. 한편 `destory` 메소드를 실행한 환경은 context라고 생각해보면, 이 테스트는 `getDeleted`가 주인공인 테스트가 되어도 괜찮을 거에요.

SUBIN606 replied:
고쳐보긴 했는데 제대로 이해한 건지 잘 모르겠네요..!

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/users/User.java:
```diff
@@ -64,4 +69,10 @@ public boolean authenticate(String rawPassword, PasswordEncoder passwordEncoder)
         return passwordEncoder.matches(rawPassword, this.password);
     }
 
+    /** 회원 삭제 시 삭제 여부를 변경한다. */
+    public User destroy() {
+        this.deleted = Boolean.TRUE;
+        return this;
+    }
```
이렇게 설명해도 괜찮겠죠.
`이 회원을 삭제 상태로 변경하고, 변경된 회원을 리턴합니다.`

johngrib reviewed (changes requested):
`getDeleted` 메소드의 이름이 `isDeleted`이면 더 좋을 것 같습니다.

johngrib commented:
코드 리뷰의 결과로 change request를 남겨 보았습니다. change request를 남기면 이렇게 빨간 색 아이콘과 함께 리뷰어의 요구를 확인할 수 있어요. 사실 이 메소드는 지금도 괜찮지만, 이 PR에 추가 작업을 하는 연습을 위해 남겨 보았습니다.

![image](https://user-images.githubusercontent.com/1855714/167131095-fa03d419-437d-49b5-a900-fb521e3277d1.png)

SUBIN606 committed:
User 엔티티 클래스 수정

- 롬복과 is~ 메서드를 함께 사용할 경우 에러가 발생할 수 있으므로 User 클래스의 `@Getter`를 제거하고 getter 메서드를 직접 작성합니다.
- getDeleted -> isDeleted 로 메서드 명을 변경합니다.
- destroy 메서드의 javaDoc을 수정합니다.

SUBIN606 committed:
테스트 코드 수정

- User 엔티티 클래스의 getDeleted -> isDeleted 메서드 명 변경에 따라 수정합니다.
- isDeleted메서드 테스트 코드를 작성합니다.

johngrib reviewed (commented):
디테일한 설명이 좋습니다. 다만 주제 하나를 콕 집어서 강조하는 것도 좋을 것 같아요. 이 작업의 주인공은 `getDeleted`를 `isDeleted`로 바꾼 것이고, 나머지 작업들은 메소드 이름 변경의 서브 작업들이라 할 수 있겠죠. 그러니 커밋 메시지도 주인공 작업을 명확히 드러내는 것이었다면 더 좋았을 거에요.

![image](https://user-images.githubusercontent.com/1855714/167255784-b0e913d1-b545-420a-8b4f-af18d516094e.png)

johngrib reviewed (approved):
좀 더 읽기 좋은 테스트 코드가 되었습니다. 고생하셨습니다!
